tools/validators/instance_validator/build/lib/validate/entity_instance.py

```diff
@@ -0,0 +1,402 @@
+
+    Args:
+      universe: ConfigUniverse generated from the ontology
+      entity_instances: dictionary containing all instances
+
+    Returns:
+      Returns boolean for validity of links key, defaulting to True if the
+      key is not present.
+    """
+
+    if self.links is None:
+      return True
+
+    is_valid = True
+
+    entity_type = universe.GetEntityType(self.namespace, self.type_name)
+    type_fields = entity_type.GetAllFields()
+    found_fields = set()
+
+    for link in self.links:
+      if link.source not in entity_instances.keys():
+        print('Invalid link source entity name: {0}'.format(link.source))
+        is_valid = False
+        continue
+
+      src_entity = entity_instances.get(link.source)
+      src_namespace = src_entity.namespace
+      src_type_name = src_entity.type_name
+      src_entity_type = universe.GetEntityType(src_namespace,
+                                                    src_type_name)
+
+      for source_field, target_field in link.field_map.items():
+        # assumes that the namespace is '' for now
+        if not entity_type.HasField('/' + target_field):
+          print('Invalid link target field: ', target_field)
+          is_valid = False
+          continue
+
+        if not src_entity_type.HasField('/' + source_field):
+          print('Invalid link source field: ', source_field)
+          is_valid = False
+          continue
+
+        found_fields.add('/' + target_field)
+
+        if not self._ValidateLinkUnits(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+        if not self._ValidateLinkStates(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+    for field_name, field in type_fields.items():
+      if not field.optional and field_name not in found_fields:
+        print('Required field {0} is missing from links'.format(field_name))
+        is_valid = False
+
+    return is_valid
+
+
+  def _ValidateLinkUnits(self, universe, source_field, target_field):
+    """Validates that units match between linked source and target fields."""
+
+    source_units = universe.GetUnitsMapByMeasurement(source_field)
+    target_units = universe.GetUnitsMapByMeasurement(target_field)
+    if source_units != target_units:
+      print('Unit mismatch in link from {0} to {1}'\
+            .format(source_field, target_field))
+      return False
+    return True
+
+
+  def _ValidateLinkStates(self, universe, source_field, target_field):
+    """Validates that states match between linked source and target fields."""
+
+    source_states = universe.GetStatesByField(source_field)
+    target_states = universe.GetStatesByField(target_field)
+    if source_states != target_states:
+      print('State mismatch in link from {0} to {1}'\
+            .format(source_field, target_field))
+      return False
+    return True
+
+
+  def IsValidEntityInstance(self, universe = None, entity_instances = None):
```

Comment:
should we assert the universe and entity instances are not none?

Reply:
We use entity_instances = None in several of the tests where it doesn't matter (because links are irrelevant).

I don't have a use case in mind for why one would want to, but the behavior when universe = None is well-defined. It will validate the entity's id and type string format only, which can be validated without the ontology.

Keeping it like this for now keeps the behavior consistent with the previous version (though the parameters used to be passed to the EntityInstance constructor instead of this validation method).

Reply:
Looking at it again I realized universe = None does lead to an attempt to call a method on None. So I added a check to terminate the validation early in that case.

Reply:
thx